src/kudu/consensus/leader_election.cc

```diff
@@ -307,34 +328,417 @@ Status FlexibleVoteCounter::RegisterVote(
+  for ( ; quorum_it != resolved_quorum.end(); quorum_it++) {
+    const std::string& region = quorum_it->first;
+    int votes_received = 0;
+    const RegionToVoterSet::const_iterator& rtvs_it =
+        region_to_voter_set.find(region);
+    if (rtvs_it != region_to_voter_set.end()) {
+      votes_received = rtvs_it->second.size();
+    }
+    int commit_requirement = quorum_it->second;
+    int votes_remaining = FetchVotesRemainingInRegion(region);
+    VLOG_WITH_PREFIX(3) << "Region: " << region << " , Votes granted: "
+                        << votes_received << " , Votes remaining: "
+                        << votes_remaining << " , Commit Requirement: "
+                        << commit_requirement;
+    if (votes_received < commit_requirement) {
+      quorum_satisfied = false;
+    }
+    if (votes_received + votes_remaining < commit_requirement) {
+      quorum_satisfaction_possible = false;
+    }
+  }
+  return std::make_pair<>(quorum_satisfied, quorum_satisfaction_possible);
+}
+
+void FlexibleVoteCounter::ConstructRegionWiseVoteCollation(
+    const int64_t term,
+    const std::string& leader_region,
+    VoteHistoryCollation* vote_collation,
+    VoteHistoryIteratorMap* it_map,
+    int64_t* min_term) const {
+  CHECK(vote_collation);
+  CHECK(it_map);
+  CHECK(min_term);
+
+  vote_collation->clear();
+  it_map->clear();
+  *min_term = 0;
+
+  VoteMap::const_iterator it = votes_.begin();
+  while (it != votes_.end()) {
+    const std::string& uuid = it->first;
+
+    // Skip servers that are not in the region of the last known leader or
+    // one of the regions that feature in the leader election quorum.
+    std::map<std::string, std::string>::const_iterator reg_it =
+        uuid_to_region_.find(uuid);
+    if (reg_it == uuid_to_region_.end() ||
+        reg_it->second != leader_region ||
+        le_quorum_requirement_.find(reg_it->second) ==
+            le_quorum_requirement_.end()) {
+      continue;
+    }
+    const std::vector<PreviousVotePB>& pvh = it->second.previous_vote_history;
+
+    // Find the voting record immediately after the term of the last known
+    // leader. Skip if there is no history beyond the last known leader.
+    std::vector<PreviousVotePB>::const_iterator vhi = std::upper_bound(
+        pvh.begin(), pvh.end(), term, compareTerm_PreviousVotePB);
+    if (vhi == pvh.end()) {
+      continue;
+    }
+    const UUIDTermPair utp = std::make_pair<>(
+        vhi->candidate_uuid(), vhi->election_term());
+
+    // Update minimum term seen so far in the last known leader's region.
+    if (reg_it->second == leader_region) {
+      *min_term = std::min(*min_term, utp.second);
+    }
+
+    // Insert the iterator into the map and update the collation.
+    // The collation is a map from (UUID, term) -> [region -> set(UUID)].
+    // For each key (UUID - term pair), it represents all servers
+    // (corresponding UUIDs) which voted for the key.
+    InsertIfNotPresent(it_map, uuid, vhi);
```

Comment:
Interesting... but the iterator to a vector is just an index. How about we just store the integer index into the vote history instead of this funky iterator?